### DISCUSSION TITLE
Vercel serves stale cached content at random

### DISCUSSION THREAD

**Discussion by willemliufdmg** [vercel/vercel#8746] [Help] [answered]
I have a Next.js application deployed on Vercel which has statically generated pages (SSG) and those pages are revalidated with the [On-Demand Revalidation](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#on-demand-revalidation) feature of Next.js.

And it seems that when a page has been revalidated Vercel can serve out both the old content and the newly revalidated content. My expectation is that only the latest content would be served.

**Reproduction:**

Demo: https://static-next.vercel.app/ssg/static-generated

Click the "Regenerate site" a bit lower on the page to trigger an On-Demand Revalidation.

You can run below script with `node etag.mjs` which does 20 requests to the demo page and collects all the unique ETAGs found.
Alternatively you can also reload the page in your webbrowser several times and you can see that sometimes the content in the textarea at the bottom of the page will change.

`etag.mjs`
```
import { setTimeout } from "timers/promises";
import { request } from "https";
const domain = "static-next.vercel.app";
const path = "/ssg/static-generated";
let numberOfRequests = 20;
const fetchEtag = () => {
  return new Promise((resolve, reject) => {
    const options = {
      host: domain,
      port: 443,
      path,
    };
    const req = request(options, (res) => {
      if (res.statusCode >= 200 && res.statusCode < 400) {
        resolve(res.headers["etag"]);
      } else {
        reject(`rejected ${res.statusCode}`);
      }
    });
    req.end();
  });
};
const results = new Set();
let n = 0;
console.log(`Sending ${n} requests to ${domain}`);
while (n < numberOfRequests) {
  console.log(`request ${n}`);
  const etag = await fetchEtag();
  if (!results.has(etag)) {
    console.log("New Key found", etag);
  }
  results.add(etag);
  n++;
  await setTimeout(100);
}
console.log(
  `Number of results found: ${results.size} in ${numberOfRequests} requests`,
  Array.from(results)
);
```

**Comment by amyegan** [comment, marked as answer]
Thanks for reaching out Willem! I don't know enough about on-demand revalidation to be able to offer a great answer myself, but I'm asking the team to take a look.

**Reply by willemliufdmg** [reply]
Our production website https://www.bnr.nl is currently suffering from this behaviour. We hope the cause of this issue can be found and fixed. Many thanks in advance!

**Reply by amyegan** [reply]
If you aren't already using the latest version of Next.js ([v12.3.1](https://www.npmjs.com/package/next/v/12.3.1)), does updating the package solve the problem?

**Reply by willemliufdmg** [reply]
I've tried that first and can confirm that it doesn't fix the problem.

**Reply by willemliufdmg** [reply]
It seems the issue has been resolved now. But I don't know what has changed in the meanwhile at Vercel. Our product hasn't changed.

**Reply by amyegan** [reply]
Sorry for the delayed response. We did some digging and found a couple of similar reports. The engineering team identified a bug that seems to be related to this issue. That bug has been addressed, but I believe they are still investigating to be sure that was the only problem. 

I'm glad that it is working for you now!
